activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb

```diff
@@ -384,7 +384,7 @@ def last_inserted_id(result)
 
         def binds_from_relation(relation, binds)
           if relation.is_a?(Relation) && binds.empty?
-            relation, binds = relation.arel, relation.bind_values
+            relation, binds = relation.arel, relation.bound_attributes
```

Comment:
`bound_attributes` isn't a concept that should exist in 4.2. Just inline the method.